feat: add progress bars to `EvaluationAgent`

DevManpreet5 opened the pull request:
## Description
This PR adds `tqdm` progress bars to metrics that require running inference, providing visual feedback during long-running evaluation operations.

## Related Issue
Fixes #336

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Changes Made
- **TotalTimeMetric (`metric_elapsed_time.py`)**: Added progress bar for inference time measurement iterations.
- **EnergyConsumedMetric (`metric_energy.py`)**: Added progress bars for both warmup and energy measurement phases.
- **EvaluationAgent (`evaluation_agent.py`)**: Added progress bar for batch processing during stateful metric evaluation.

## How Has This Been Tested?
- Verified that progress bars appear during evaluation testcases.

![Batch processing](https://github.com/user-attachments/assets/f37993f3-0609-41f8-be1b-0ac4b5c5eaeb)  
*Figure 1: Progress bar during batch processing in EvaluationAgent.*  

`uv run pytest tests/evaluation/test_cmmd.py -v -s --no-cov`

![Warmup & Energy](https://github.com/user-attachments/assets/da7aeae2-590a-4172-a505-c2007a62812a)  
*Figure 2: Progress bars during warmup and energy measurement.*  

`uv run pytest tests/evaluation/test_energy_metrics.py -v -s --no-cov`

![Inference time](https://github.com/user-attachments/assets/a0bcd0a4-2cdc-4e49-8236-be35f89d0db9)  
*Figure 3: Progress bar for inference time measurement.*  

`uv run pytest tests/evaluation/test_inference_time_metrics.py::test_total_time_metric -v -s --capture=no`

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature works
- [x] New and existing unit tests pass locally with my changes

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds tqdm progress bars for batch processing and for inference timing and energy measurement loops during evaluation.
> 
> - **Evaluation**:
>   - `EvaluationAgent.update_stateful_metrics`: Wraps dataloader loop with `tqdm` to show batch processing progress.
> - **Metrics**:
>   - `metric_elapsed_time.py` (`InferenceTimeStats`): Adds `tqdm` progress bar for measured iterations during inference timing.
>   - `metric_energy.py` (`EnvironmentalImpactStats`): Adds `tqdm` progress bars for warmup and inference phases when tracking energy/CO2.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit aab5a039a37225cd908323cfc83cb4fd2c633584. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

DevManpreet5 commented:
@davidberenstein1957 Can you pls review it for #336

johannaSommer commented on src/pruna/evaluation/evaluation_agent.py:
```diff
@@ -200,7 +201,7 @@ def update_stateful_metrics(
 
         model.move_to_device(self.device)
-        for batch_idx, batch in enumerate(self.task.dataloader):
+        for batch_idx, batch in enumerate(tqdm(self.task.dataloader, desc="Processing batchs", unit="batch")):
```
typo: *batches

johannaSommer commented on src/pruna/evaluation/metrics/metric_energy.py:
```diff
@@ -104,11 +105,11 @@ def compute(self, model: PrunaModel, dataloader: DataLoader) -> Dict[str, Any] |
 
         # Warmup
-        for _ in range(self.n_warmup_iterations):
+        for _ in tqdm(range(self.n_warmup_iterations), desc="Warmup", unit="iter"):
```
Can we extend the description here to by slightly more specific? E.g. "Warm-up for energy consumption metric" or something along those lines?

begumcig replied:
+1

johannaSommer reviewed (changes requested):
I love the changes, thanks a ton @DevManpreet5 !! 🌻 Just a couple of comments about the naming on the progress bars, otherwise it's already G2G. One question - does the user have a simple way of disabling the progress bars, e.g. via a tqdm environment variable? I feel like this would be useful to introduce as well.

begumcig reviewed (changes requested):
Wow this is an insanely cool feature thanks a lot for implementing @DevManpreet5 !!! Just the small typos / improving log messages but looks super good otherwise!

DevManpreet5 commented:
@johannaSommer @begumcig Thanks for taking the time to review the PR!

1. I’ve fixed the typo and updated the description for warmup energy.

2.  > One question - does the user have a simple way of disabling the progress bars, e.g. via a tqdm environment variable?

 Regarding the question about disabling progress bars via the tqdm environment variable: users can already do this using the standard TQDM_DISABLE=1.

Attaching images to show the difference between the normal behavior and when TQDM_DISABLE=1 is set.

Original `uv run pytest tests/evaluation/test_energy_metrics.py -v -s --no-cov` -> progress bars enabled:

<img width="1434" height="613" alt="normal" src="https://github.com/user-attachments/assets/2136c4b9-ae51-4964-8b8e-bfd2177c6d94" />

With disable `TQDM_DISABLE=1 uv run pytest tests/evaluation/test_energy_metrics.py -v -s --no-cov ` -> progress bars disabled:
 
<img width="1424" height="601" alt="disable" src="https://github.com/user-attachments/assets/cca9788d-0c8e-4a17-b7ed-7c07795b525f" />

begumcig reviewed (approved):
Thank you for the updates @DevManpreet5! Looks super good to me!
